### PR TITLE
Propagate types through lambdas

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

File: src/compiler/propagation/type_propagator.cc [hunk 1]
```diff
@@ -294,6 +294,17 @@ void TypePropagator::call_virtual(MethodTemplate* caller, TypeStack* stack, uint
   stack->drop_arguments(arity);
 }
 
+void TypePropagator::propagate_through_lambda(Method method) {
+  ASSERT(method.is_lambda_method());
+  std::vector<ConcreteType> arguments;
+  // TODO(kasper): Can we at least push an instance of the Lambda class
+  // as the receiver type?
+  for (int i = 0; i < method.arity(); i++) {
+    arguments.push_back(ConcreteType::any());
```
Comment (floitsch) [inline src/compiler/propagation/type_propagator.cc:303]: Why can't we use the types that are on the stack?
Add a comment.

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/compiler/propagation/type_propagator.cc [hunk 2]
```diff
@@ -620,16 +631,21 @@ static void process(MethodTemplate* method, uint8* bcp, TypeScope* scope, Workli
...
   OPCODE_END();
 
-  OPCODE_BEGIN(LOAD_BLOCK_METHOD);
+  OPCODE_BEGIN(LOAD_METHOD);
     Method inner = Method(program->bytecodes, Utils::read_unaligned_uint32(bcp + 1));
     // Finds or creates a block-template for the given block.
     // The block's parameters are marked such that a change in their type enqueues this
     // current method template.
     // Note that the method template is for a specific combination of parameter types. As
     // such we evaluate the contained blocks independently too.
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/compiler/propagation/type_propagator.cc]: Done.

File: src/compiler/propagation/type_propagator.cc (hunk 1 above)
Comment (kasperl) [reply src/compiler/propagation/type_propagator.cc:319]: We're instantiating -- not invoking -- the lambda here. Added comment.

**Pull request merged**